driver-core/src/test/unit/com/mongodb/internal/connection/IndexMapSpecification.groovy

```diff
@@ -81,4 +81,33 @@ class IndexMapSpecification extends Specification {
         1 == indexMap.map(0)
         2 == indexMap.map(1)
     }
-}
\ No newline at end of file
+
+    def 'should include ranges when converting from range based to hash based indexMap'() {
+        given:
+        def indexMap = IndexMap.create(1000, 3)
+
+        when: 'converts from range based with a high startIndex to hash based'
+        indexMap = indexMap.add(5, 1005)
+
+        then:
+        1000 == indexMap.map(0)
```

Comment:
To test that loop condition better, let's assert that indexMap.map(-1) and indexMap.map(3) both throw MongoInternalException